### PR TITLE
fix(pin): add select for channel handling

### PR DESCRIPTION
**Pull request by overbool**
License: MIT
Signed-off-by: Overbool <overbool.xu@gmail.com>

### CODE REVIEW

File: core/commands/pin.go [hunk 1]
```diff
@@ -115,12 +118,31 @@ var addPinCmd = &cmds.Command{
 				}
 
 				if pv := v.Value(); pv != 0 {
-					out <- &AddPinOutput{Progress: v.Value()}
+					select {
+					case out <- &AddPinOutput{Progress: v.Value()}:
+					case <-ctx.Done():
+						log.Error(ctx.Err())
```
Comment (magik6k) [inline core/commands/pin.go]: This will print on the daemon, pretty sure we don't want that

Comment (kevina) [reply core/commands/pin.go]: Agree.  Just return the error.

**Comment by kevina** [comment]
> I'd wait for #5035 to land and just rewrite it to the new command lib / res.Emit instead of messing witch channels here

That seams like an overkill to me, but I am not against it.

File: core/commands/pin.go [hunk 2]
```diff
@@ -100,7 +100,10 @@ var addPinCmd = &cmds.Command{
 		ch := make(chan pinResult, 1)
 		go func() {
 			added, err := corerepo.Pin(n, ctx, req.Arguments(), recursive)
-			ch <- pinResult{pins: added, err: err}
+			select {
+			case ch <- pinResult{pins: added, err: err}:
+			case <-ctx.Done():
```
[the remark this answers is not in the document]
Comment (overbool) [reply core/commands/pin.go]: @magik6k Although `ch` is a buffered `chan`, if the `select case` of receiving value from this `chan` returns or exits(like https://github.com/ipfs/go-ipfs/blob/9bf4e4145ea897ced2f9aa2087296b3beda6e941/core/commands/pin.go#L118-L132), this goroutine will be blocked.

**Pull request closed without merging**